File: prototype/build_site.py
#!/usr/bin/env python3
# build_site.py — 读取 skills/*/SKILL.md 的 frontmatter 生成展示页数据 data/skills.json
# 路径: prototype/build_site.py 版本: 1.5.0
#
# 数据源策略：
#   主数据源为各技能 SKILL.md 的 frontmatter（name / description / category）。
#   仅当 SKILL.md 缺 category 时，回退到 README.md 的分类映射，保证兼容性。
#
# 用法：
#   python3 build_site.py            # 生成 data/skills.json
#   python3 build_site.py --check    # CI 校验：已提交的 JSON 是否与最新生成一致
#
# 注：本仓库同时提供 Node 移植版 build_site.mjs（逻辑 1:1 对齐），
#     供无 Python 环境使用；npm run build 默认调用 build_site.mjs。

import json
import os
import re
import sys

# 仓库根目录（脚本位于 prototype/ 下，上级目录即仓库根）
ROOT = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
README = os.path.join(ROOT, "README.md")
SKILLS_DIR = os.path.join(ROOT, "skills")
OUT = os.path.join(ROOT, "prototype", "data", "skills.json")

# 仓库实际地址（兜底值，README 解析失败时使用）
DEFAULT_REPO = "https://github.com/sutchan/Agent-Skills-Hub"

# README 中定义的合法分类顺序（用于排序回退）
README_CAT_ORDER = []

# 分类中文名 -> 英文显示名（与展示页原型保持一致；新增分类时在此补充）
CAT_EN = {
    "前端与 UI 设计": "Frontend & UI Design",
    "后端、语言与框架": "Backend, Languages & Frameworks",
    "架构与设计": "Architecture & Design",
    "测试与质量": "Testing & Quality",
    "Agent 与 AI 工程": "Agent & AI Engineering",
    "DevOps 与基础设施": "DevOps & Infrastructure",
    "数据与机器学习": "Data & Machine Learning",
    "内容、文档与写作": "Content, Docs & Writing",
    "视频与媒体": "Video & Media",
    "行业领域": "Industry Domains",
    "生产力与工具": "Productivity & Tools",
    "上下文与提示工程": "Context & Prompt Engineering",
    "其他": "Other",
}



def parse_readme_categories():
    """解析 README，返回 (分类顺序, 技能名->分类名, 技能名->中文描述)。"""
    cat_map = {}
    zh_desc_map = {}
    order = []
    cat_re = re.compile(r"^###\s+(.+?)(?:（(\d+)）)?$")
    item_re = re.compile(r"^\s*-\s+\*\*\[([^\]]+)\]\(([^)]+)\)\*\*\s*[—-]\s*(.+)$")
    current_cat = None
    with open(README, "r", encoding="utf-8") as f:
        lines = f.read().splitlines()
    for line in lines:
        if re.match(r"^###\s", line) and not line.startswith("## "):
            m = cat_re.match(line)
            if not m:
                continue
            name = m.group(1).strip()
            if name in ("目录", "仓库结构"):
                current_cat = None
                continue
            current_cat = name
            if current_cat not in order:
                order.append(current_cat)
            continue
        if current_cat is None:
            continue
        im = item_re.match(line)
        if im:
            skill = im.group(1).strip()
            desc = im.group(3).strip().rstrip("。 ")
            cat_map[skill] = current_cat
            zh_desc_map[skill] = desc
    return order, cat_map, zh_desc_map


def read_skill_meta(skill_dir):
    """读取 SKILL.md 的 frontmatter（name/description/category）与资源目录标记。"""
    skill_path = os.path.join(SKILLS_DIR, skill_dir)
    if not os.path.isdir(skill_path):
        return None
    skill_md = os.path.join(skill_path, "SKILL.md")
    if not os.path.isfile(skill_md):
        return None
    with open(skill_md, "r", encoding="utf-8") as f:
        content = f.read()
    meta = {
        "en_name": skill_dir,
        "en_desc": "",
        "name": skill_dir,
        "category": "",
    }
    m = re.match(r"^---\s*\n(.*?)\n---", content, re.DOTALL)
    if m:
        fm = m.group(1)
        for key in ("name", "description", "category"):
            km = re.search(rf"^{key}:\s*(.+)$", fm, re.MULTILINE)
            if km:
                val = km.group(1).strip()
                # 去除 YAML 标量常见引号包裹（"..." 或 '...'），避免 name 带引号进入数据
                if len(val) >= 2 and val[0] == val[-1] and val[0] in ("\"", "'"):
                    val = val[1:-1].strip()
                meta[key if key in ("name", "category") else "en_desc"] = val
    meta["has_scripts"] = os.path.isdir(os.path.join(skill_path, "scripts"))
    meta["has_references"] = os.path.isdir(os.path.join(skill_path, "references"))
    meta["has_assets"] = os.path.isdir(os.path.join(skill_path, "assets"))
    meta["dir"] = skill_dir
    return meta


def parse_readme_meta():
    """解析 README 顶部元信息（作者 / 仓库地址 / 总数）。"""
    meta = {"author": "", "repo": "", "count": ""}
    with open(README, "r", encoding="utf-8") as f:
        lines = f.read().splitlines()
    for line in lines[:15]:
        m = re.search(r"作者[：:]\s*(.+)", line)
        if m:
            meta["author"] = m.group(1).strip()
        m = re.search(r"项目地址[：:]\s*(\S+)", line)
        if m:
            meta["repo"] = m.group(1).strip()
    return meta


def build_data():
    readme_meta = parse_readme_meta()
    readme_order, readme_cat_map, readme_zh_map = parse_readme_categories()

    skills = []
    seen = set()
    # 遍历磁盘上所有含 SKILL.md 的技能目录（单一数据源）
    for entry in sorted(os.listdir(SKILLS_DIR)):
        skill_path = os.path.join(SKILLS_DIR, entry)
        if not os.path.isdir(skill_path) or not os.path.isfile(
            os.path.join(skill_path, "SKILL.md")
        ):
            continue
        if entry in seen:
            continue
        seen.add(entry)
        sm = read_skill_meta(entry)
        if sm is None:
            continue
        # 分类优先取 SKILL.md frontmatter，缺失则回退 README 映射
        if not sm.get("category"):
            sm["category"] = readme_cat_map.get(entry, "其他")
        # 中文描述以 README 为准（已本地化），缺失则为空
        sm["zh_desc"] = readme_zh_map.get(entry, "")
        skills.append(sm)

    # 分类顺序：优先 README 顺序，再补齐 frontmatter 中出现但 README 未列的分类
    cat_counts = {}
    for s in skills:
        cat_counts[s["category"]] = cat_counts.get(s["category"], 0) + 1
    ordered_cats = [c for c in readme_order if c in cat_counts]
    for c in cat_counts:
        if c not in ordered_cats:
            ordered_cats.append(c)

    categories = [
        {"name": c, "en": CAT_EN.get(c, c), "count": cat_counts[c]}
        for c in ordered_cats
    ]

    data = {
        "meta": {
            "title": "Agent Skills Hub",
            "subtitle": "面向开发、设计、测试、DevOps、Agent 工程及各行业领域的 AI 技能集合",
            "author": readme_meta["author"],
            "repo": readme_meta["repo"] or DEFAULT_REPO,
            "count": len(skills),
            "generated_at": __import__("datetime").datetime.now().isoformat(timespec="seconds"),
        },
        "categories": categories,
        "skills": skills,
    }
    return data


def main():
    check_mode = "--check" in sys.argv
    data = build_data()

    if check_mode:
        if not os.path.isfile(OUT):
            print("ERROR: 未找到已提交的 data/skills.json，请先运行 build_site.py 生成。")
            sys.exit(1)
        with open(OUT, "r", encoding="utf-8") as f:
            existing = json.load(f)
        # 仅比较结构性字段（忽略 generated_at 时间戳）
        def normalize(d):
            return {
                "meta": {k: v for k, v in d["meta"].items() if k != "generated_at"},
                "categories": d["categories"],
                "skills": d["skills"],
            }
        if normalize(existing) != normalize(data):
            print("ERROR: data/skills.json 与最新 skills/ 不一致，请运行 build_site.py 重新生成。")
            sys.exit(1)
        print(f"OK: data/skills.json 已是最新（{len(data['skills'])} 个技能，{len(data['categories'])} 个分类）。")
        return

    os.makedirs(os.path.dirname(OUT), exist_ok=True)
    with open(OUT, "w", encoding="utf-8") as f:
        json.dump(data, f, ensure_ascii=False, indent=2)

    print(f"已生成 {OUT}")
    print(f"技能总数: {len(data['skills'])}  分类数: {len(data['categories'])}")


if __name__ == "__main__":
    main()

File: prototype/build_site.mjs
#!/usr/bin/env node
// build_site.mjs — 读取 skills/*/SKILL.md 的 frontmatter 生成展示页数据 data/skills.json
// 路径: prototype/build_site.mjs 版本: 1.7.0
// 说明：与 build_site.py 逻辑 1:1 对齐，供无 Python 环境（如本机构建）使用。
//       npm run build 默认调用本文件；Python 环境下也可运行 build_site.py。
//
// 用法：
//   node build_site.mjs            # 生成 data/skills.json
//   node build_site.mjs --check    # CI 校验：已提交的 JSON 是否与最新生成一致

import fs from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const ROOT = path.dirname(__dirname);
const README = path.join(ROOT, "README.md");
const SKILLS_DIR = path.join(ROOT, "skills");
const OUT = path.join(ROOT, "prototype", "data", "skills.json");

const DEFAULT_REPO = "https://github.com/sutchan/Agent-Skills-Hub";

// 分类中文名 -> 英文显示名（与展示页原型保持一致；新增分类时在此补充）
const CAT_EN = {
  前端与UI设计: "Frontend & UI Design",
  "后端、语言与框架": "Backend, Languages & Frameworks",
  架构与设计: "Architecture & Design",
  测试与质量: "Testing & Quality",
  "Agent 与 AI 工程": "Agent & AI Engineering",
  "DevOps 与基础设施": "DevOps & Infrastructure",
  数据与机器学习: "Data & Machine Learning",
  "内容、文档与写作": "Content, Docs & Writing",
  视频与媒体: "Video & Media",
  行业领域: "Industry Domains",
  生产力与工具: "Productivity & Tools",
  上下文与提示工程: "Context & Prompt Engineering",
  其他: "Other",
};

function stripQuotes(val) {
  const s = val.trim();
  if (s.length >= 2 && s[0] === s[s.length - 1] && (s[0] === '"' || s[0] === "'")) {
    return s.slice(1, -1).trim();
  }
  return s;
}

function parseReadme() {
  const text = fs.readFileSync(README, "utf-8");
  const lines = text.split(/\r?\n/);
  const catMap = {};
  const zhDescMap = {};
  const order = [];
  let currentCat = null;
  const catRe = /^###\s+(.+?)(?:（(\d+)）)?$/;
  const itemRe = /^\s*-\s+\*\*\[([^\]]+)\]\(([^)]+)\)\*\*\s*[—-]\s*(.+)$/;
  for (const line of lines) {
    if (line.startsWith("### ") && !line.startsWith("## ")) {
      const m = catRe.exec(line);
      if (!m) continue;
      const name = m[1].trim();
      if (name === "目录" || name === "仓库结构") {
        currentCat = null;
        continue;
      }
      currentCat = name;
      if (!order.includes(currentCat)) order.push(currentCat);
      continue;
    }
    if (currentCat === null) continue;
    const im = itemRe.exec(line);
    if (im) {
      const skill = im[1].trim();
      const desc = im[3].trim().replace(/[。\s]+$/, "");
      catMap[skill] = currentCat;
      zhDescMap[skill] = desc;
    }
  }
  return { order, catMap, zhDescMap };
}

function readSkillMeta(skillDir) {
  const skillPath = path.join(SKILLS_DIR, skillDir);
  if (!fs.statSync(skillPath, { throwIfNoEntry: false })?.isDirectory())
    return null;
  const skillMd = path.join(skillPath, "SKILL.md");
  if (!fs.existsSync(skillMd)) return null;
  const content = fs.readFileSync(skillMd, "utf-8");
  const meta = {
    en_name: skillDir,
    en_desc: "",
    name: skillDir,
    category: "",
  };
  const m = /^---\s*\n([\s\S]*?)\n---/.exec(content);
  if (m) {
    const fm = m[1];
    for (const key of ["name", "description", "category"]) {
      const km = new RegExp(`^${key}:\\s*(.+)$`, "m").exec(fm);
      if (km) {
        const val = stripQuotes(km[1]);
        meta[key === "description" ? "en_desc" : key] = val;
      }
    }
  }
  meta.has_scripts = fs.existsSync(path.join(skillPath, "scripts"));
  meta.has_references = fs.existsSync(path.join(skillPath, "references"));
  meta.has_assets = fs.existsSync(path.join(skillPath, "assets"));
  meta.dir = skillDir;
  return meta;
}

function parseReadmeMeta() {
  const lines = fs.readFileSync(README, "utf-8").split(/\r?\n/).slice(0, 15);
  const meta = { author: "", repo: "" };
  for (const line of lines) {
    const am = /作者[：:]\s*(.+)/.exec(line);
    if (am) meta.author = am[1].trim();
    const rm = /项目地址[：:]\s*(\S+)/.exec(line);
    if (rm) meta.repo = rm[1].trim();
  }
  return meta;
}

function buildData() {
  const readmeMeta = parseReadmeMeta();
  const { order, catMap, zhDescMap } = parseReadme();

  const skills = [];
  const seen = new Set();
  for (const entry of fs.readdirSync(SKILLS_DIR).sort()) {
    const skillPath = path.join(SKILLS_DIR, entry);
    if (!fs.statSync(skillPath, { throwIfNoEntry: false })?.isDirectory())
      continue;
    if (!fs.existsSync(path.join(skillPath, "SKILL.md"))) continue;
    if (seen.has(entry)) continue;
    seen.add(entry);
    const sm = readSkillMeta(entry);
    if (!sm) continue;
    if (!sm.category) sm.category = catMap[entry] || "其他";
    sm.zh_desc = zhDescMap[entry] || "";
    skills.push(sm);
  }

  const catCounts = {};
  for (const s of skills)
    catCounts[s.category] = (catCounts[s.category] || 0) + 1;
  const orderedCats = order.filter((c) => c in catCounts);
  for (const c of Object.keys(catCounts)) if (!orderedCats.includes(c)) orderedCats.push(c);

  const categories = orderedCats.map((c) => ({
    name: c,
    en: CAT_EN[c] || c,
    count: catCounts[c],
  }));

  return {
    meta: {
      title: "Agent Skills Hub",
      title_zh: "Agent Skills Hub · 技能中心",
      subtitle: "面向开发、设计、测试、DevOps、Agent 工程及各行业领域的 AI 技能集合",
      author: readmeMeta.author,
      repo: readmeMeta.repo || DEFAULT_REPO,
      count: skills.length,
      generated_at: new Date().toISOString().replace(/\.\d+Z$/, ""),
    },
    categories,
    skills,
  };
}

function normalize(d) {
  return {
    meta: Object.fromEntries(
      Object.entries(d.meta).filter(([k]) => k !== "generated_at")
    ),
    categories: d.categories,
    skills: d.skills,
  };
}

function main() {
  const checkMode = process.argv.includes("--check");
  const data = buildData();

  if (checkMode) {
    if (!fs.existsSync(OUT)) {
      console.error("ERROR: 未找到已提交的 data/skills.json，请先运行 build_site.mjs 生成。");
      process.exit(1);
    }
    const existing = JSON.parse(fs.readFileSync(OUT, "utf-8"));
    if (JSON.stringify(normalize(existing)) !== JSON.stringify(normalize(data))) {
      console.error("ERROR: data/skills.json 与最新 skills/ 不一致，请运行 build_site.mjs 重新生成。");
      process.exit(1);
    }
    console.log(
      `OK: data/skills.json 已是最新（${data.skills.length} 个技能，${data.categories.length} 个分类）。`
    );
    return;
  }

  fs.mkdirSync(path.dirname(OUT), { recursive: true });
  fs.writeFileSync(OUT, JSON.stringify(data, null, 2), "utf-8");
  console.log(`已生成 ${OUT}`);
  console.log(`技能总数: ${data.skills.length}  分类数: ${data.categories.length}`);
}

main();
